src/composer/compiler.rs

```diff
@@ -0,0 +1,388 @@
+            q_d[i] = c.q_d;
+            q_arith[i] = c.q_arith;
+            q_range[i] = c.q_range;
+            q_logic[i] = c.q_logic;
+            q_fixed_group_add[i] = c.q_fixed_group_add;
+            q_variable_group_add[i] = c.q_variable_group_add;
+        });
+
+        let q_m_poly = domain.ifft(&q_m);
+        let q_l_poly = domain.ifft(&q_l);
+        let q_r_poly = domain.ifft(&q_r);
+        let q_o_poly = domain.ifft(&q_o);
+        let q_c_poly = domain.ifft(&q_c);
+        let q_d_poly = domain.ifft(&q_d);
+        let q_arith_poly = domain.ifft(&q_arith);
+        let q_range_poly = domain.ifft(&q_range);
+        let q_logic_poly = domain.ifft(&q_logic);
+        let q_fixed_group_add_poly = domain.ifft(&q_fixed_group_add);
+        let q_variable_group_add_poly = domain.ifft(&q_variable_group_add);
+
+        let q_m_poly = FftPolynomial::from_coefficients_vec(q_m_poly);
+        let q_l_poly = FftPolynomial::from_coefficients_vec(q_l_poly);
+        let q_r_poly = FftPolynomial::from_coefficients_vec(q_r_poly);
+        let q_o_poly = FftPolynomial::from_coefficients_vec(q_o_poly);
+        let q_c_poly = FftPolynomial::from_coefficients_vec(q_c_poly);
+        let q_d_poly = FftPolynomial::from_coefficients_vec(q_d_poly);
+        let q_arith_poly = FftPolynomial::from_coefficients_vec(q_arith_poly);
+        let q_range_poly = FftPolynomial::from_coefficients_vec(q_range_poly);
+        let q_logic_poly = FftPolynomial::from_coefficients_vec(q_logic_poly);
+        let q_fixed_group_add_poly =
+            FftPolynomial::from_coefficients_vec(q_fixed_group_add_poly);
+        let q_variable_group_add_poly =
+            FftPolynomial::from_coefficients_vec(q_variable_group_add_poly);
+
+        // 2. compute the sigma polynomials
+        let [s_sigma_1_poly, s_sigma_2_poly, s_sigma_3_poly, s_sigma_4_poly] =
+            perm.compute_sigma_polynomials(size, &domain);
+
+        let q_m_poly_commit = commit_key.commit(&q_m_poly).unwrap_or_default();
+        let q_l_poly_commit = commit_key.commit(&q_l_poly).unwrap_or_default();
+        let q_r_poly_commit = commit_key.commit(&q_r_poly).unwrap_or_default();
+        let q_o_poly_commit = commit_key.commit(&q_o_poly).unwrap_or_default();
+        let q_c_poly_commit = commit_key.commit(&q_c_poly).unwrap_or_default();
+        let q_d_poly_commit = commit_key.commit(&q_d_poly).unwrap_or_default();
+        let q_arith_poly_commit =
+            commit_key.commit(&q_arith_poly).unwrap_or_default();
+        let q_range_poly_commit =
+            commit_key.commit(&q_range_poly).unwrap_or_default();
+        let q_logic_poly_commit =
+            commit_key.commit(&q_logic_poly).unwrap_or_default();
+        let q_fixed_group_add_poly_commit = commit_key
+            .commit(&q_fixed_group_add_poly)
+            .unwrap_or_default();
+        let q_variable_group_add_poly_commit = commit_key
+            .commit(&q_variable_group_add_poly)
+            .unwrap_or_default();
```

Comment:
This is just a multi-scalar multiplication, and shouldn't be fallible. For some reason, these commits are re-checking the commit degree - that was already checked. This is perfectly safe since it won't compromise the validity of the verification.

It was not the intent of this PR to fix the problems of the KZG implementation - but just to shield the API from leaking them. It doesn't make sense to extend this redundant check to the composer/prover, so its implemented that way.

I added a comment with a TODO to fix that in KZG